Домашнее задание 11 спринта

DaniilKosmachev opened the pull request.

DaniilKosmachev committed:
ДЗ 11 спринта

DaniilKosmachev committed:
ДЗ 11 спринта

DaniilKosmachev committed:
ДЗ 11 спринта

DaniilKosmachev committed:
ДЗ 11 спринта

DaniilKosmachev committed:
Merge pull request #4 from DaniilKosmachev/testDev

Test dev

DaniilKosmachev committed:
ДЗ 11 спринта, вариант 1:
1) Реализовал экраны (пустой ответ и проблемы с коннектом) путем разных viewType в одном RecyclerView

DaniilKosmachev committed:
ДЗ 11 спринта, вариант 2:
1) Реализовал экраны (пустой ответ и проблемы с коннектом) путем разных viewType в одном RecyclerView
1.1) Чуть переименовал интерфейс, поставил комменты

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/EmptySearchViewHolder.kt:
```diff
@@ -0,0 +1,10 @@
+
+class EmptySearchPlaceHolder : TrackAdapterListItem
+
+class EmptySearchViewHolder(parent: ViewGroup): ViewHolder(
```
⚠️Лучше не реализовывать плейсхолдер как элемент списка. Главная проблема заключается в сохранении анимации скролла. 
![Screenshot_20230903_131733](https://github.com/DaniilKosmachev/PlaylistMaker/assets/57040048/b607cb8c-bd7c-4789-a2da-4d3f6cc53ad0)

DaniilKosmachev replied:
То есть лучше не делать элементами RecyclerView? А скрывать RecyclerView и отображать ошибки?

nasibullin1412 replied:
Да

DaniilKosmachev replied:
Понял, сделаю)) спасибо)

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/MainActivity.kt:
```diff
@@ -13,14 +13,12 @@ class MainActivity : AppCompatActivity(), View.OnClickListener {
     override fun onCreate(savedInstanceState: Bundle?) {
```
🍏Советую не засорять `onCreate` и разнести весь код отсюда по функциям, которые будут вызываться из `onCreate`

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/NoConnectionViewHolder.kt:
```diff
@@ -0,0 +1,15 @@
+
+class NoConnectionPlaceHolder: TrackAdapterListItem
+
+class NoConnectionViewHolder(parent: ViewGroup): ViewHolder(
```
⚠️Лучше не реализовывать плейсхолдер как элемент списка. Главная проблема заключается в сохранении анимации скролла.

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -7,13 +7,29 @@ import android.os.Bundle
 import android.widget.EditText
 import android.widget.ImageButton
 import androidx.recyclerview.widget.RecyclerView
+import retrofit2.*
```
🍏Старайся не использовать `wildcard import`

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -7,13 +7,29 @@ import android.os.Bundle
+import retrofit2.converter.gson.GsonConverterFactory
+
+class SearchActivity : AppCompatActivity(), TrackAdapter.OnUpdateButtonClickListener {//реализуем интерфейс, написанный в TrackAdapter
+    private val iTunesBaseUrl = "https://itunes.apple.com"
```
🍏 Неизменяемый базовый тип лучше сразу выносить в `companion object `и делать `const val`. Отличие `val `от `const val` описано здесь https://tech-geek.ru/val-const-val-kotlin/

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -67,17 +87,64 @@ class SearchActivity : AppCompatActivity() {
+                    startSearchTrack()//запускаем метод поиска треков
+                }
+                true
+            } else
+            {
```
🍏 Лишний перенос строки

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -67,17 +87,64 @@ class SearchActivity : AppCompatActivity() {
+        }
+    }
+    private fun startSearchTrack() {
+        val inputMethodManager = getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
+        inputMethodManager?.hideSoftInputFromWindow(editTextSearch.windowToken,0)//сразу сворачиваем клавиатуру - не удобно с ней
```
🍏Эта логика никак не относится к поиску трека. Лучше вынести её за пределы функции `startSearchTrack`

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/TrackAdapterListItem.kt:
```diff
@@ -0,0 +1,4 @@
+package com.example.playlistmaker
+
+interface TrackAdapterListItem {
```
🍏Можно оставить просто `interface TrackAdapterListItem`

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/TracksResponse.kt:
```diff
@@ -0,0 +1,4 @@
+package com.example.playlistmaker
+
+class TracksResponse (val resultCount: Int, val results: List<Track>) {
```
🍏Можно оставить просто `class TracksResponse (val resultCount: Int, val results: List<Track>)` без скобок

nasibullin1412 commented on app/src/main/res/layout/empty_search_view.xml:
```diff
@@ -0,0 +1,14 @@
+    android:orientation="vertical"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:gravity="center">
```
⚠️Плейсхолдер  в фигме не находится в центре

DaniilKosmachev committed:
ДЗ 11 спринта, вариант 3:
1) Реализовал плейсхолдеры по-другому (визибл и т.п.)
2) Почистил методы onCreate всех активностей, вынеся лишнее в методы
3) Убрал wildcard import из SearchActivity.kt
4) Base Url стал const val в companion objcet
5) Создал отдельный метод для сворачивания системной клавиатуры
6) Теперь плейсхолдеры (LinearLayout) не по центру сам, а только элементы, которые содержаться в нем
7) Удалил лишние классы, так как теперь другая логика

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -31,28 +64,35 @@ class SearchActivity : AppCompatActivity() {
-        buttonBack.setOnClickListener {
-            finish()
-        }
+        buttonBack = findViewById(R.id.backToMainActivityViewButtonFromSearch)
+        editTextSearch = findViewById(R.id.editTextSearchActivity)
+        clearButton = findViewById(R.id.clearEditTextSearchActivity)
+        recycleViewTrack = findViewById(R.id.track_recycle_view)
+        updateButtonClick = findViewById(R.id.updateQueryButton)
+        errorLayout = findViewById(R.id.search_activity_error_linear_layout)
+        errorImage = findViewById(R.id.search_activity_error_image)
+        errorText = findViewById(R.id.search_activity_error_text)
```
🍏 Это тоже лучше вынести в отдельную функцию

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/SearchActivity.kt:
```diff
@@ -68,16 +108,134 @@ class SearchActivity : AppCompatActivity() {
-        val recycleViewTrack = findViewById<RecyclerView>(R.id.track_recycle_view)
-        recycleViewTrack.adapter = trackAdapter
+    }
+    @SuppressLint("NotifyDataSetChanged")
```
🍏Оставляй пустую строку между методами

nasibullin1412 commented on app/src/main/java/com/example/playlistmaker/TrackViewHolder.kt:
```diff
@@ -1,14 +1,13 @@
 import androidx.recyclerview.widget.RecyclerView
 import com.bumptech.glide.Glide
 import com.bumptech.glide.load.resource.bitmap.RoundedCorners
+import java.util.*
```
🍏`wildcard import`, советую прочитать статью, чтобы узнать как их отключить в студии https://www.baeldung.com/intellij-disable-wildcard-import

DaniilKosmachev committed:
ДЗ 11 спринта, вариант 4:
1) Сделал строки между методами
2) Вынес инициализацию вьюшек в отдельный метод, вызываю его в onCreate
3) Убрал импорт *
